### PR TITLE
User can create account

### PR DESCRIPTION
**Pull request by Carrosen**
# PT Story

## Description
Changes proposed in this pull request:
* Adding a Feature test for "User can create account" (8 steps)
* Adding step definitions (one new file for assertions)
* Adding unit tests (RSpec) for saving user: name, email and password to database and validating their presence

## What I have learned working on this feature:
* Capybara syntax
* Cucumber sometimes giving false positives
* Better understanding of permitted params for sign up functionality

## User Story
Link to Pivotal Tracker where you can find the [user story](https://www.pivotaltracker.com/story/show/165573162)

```
As a User,
In order to use the application
I would like to be able to create an account
```

## Pair Programming
* [leiter007](https://github.com/leiter007)
* [Carrosen](https://github.com/Carrosen)

## Testing frameworks
* Cucumber
* Rspec

### CODE REVIEW

File: features/step_definitions/user_can_create_account_assertions_steps.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+
+Then("I should see {string}") do |content|
```
Comment (oliverochman) [inline features/step_definitions/user_can_create_account_assertions_steps.rb]: rename this file to `assertions_steps` instead

File: features/step_definitions/user_can_create_account_steps.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (oliverochman) [inline features/step_definitions/user_can_create_account_steps.rb]: rename this file to `basic_steps` instead

File: features/step_definitions/user_can_create_account_steps.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+Given("I visit the {string} page") do |string|
+  visit root_path
```
Comment (oliverochman) [inline features/step_definitions/user_can_create_account_steps.rb]: You do not use the string here, why?

Perhaps rethink this step

File: features/user_can_create_account.feature [hunk 4]
```diff
@@ -0,0 +1,29 @@
+Feature: User can create account
+    As a User,
+    In order to use the application
+    I would like to be able to create an account
+
+    Background: 
+        Given I visit the "landing" page
```
Comment (oliverochman) [inline features/user_can_create_account.feature]: I would remove the quotes around landing page for this step.

File: features/user_can_create_account.feature [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    Scenario: Successfully create an account [Happy Path]
+        When I fill in "Name" with "Feliks"
+        And I fill in "Email" with "feliks@craft.se"
+        And I fill in "Password" with "password"
+        And I fill in "Password confirmation" with "password"
+        And I click "Create"
+        Then I should see "Welcome! You have signed up successfully"
+   
+
+        #[Sad Path]
```
Comment (oliverochman) [inline features/user_can_create_account.feature]: Please finish the Sad Path

File: spec/models/user_can_create_account_spec.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    describe "DB table" do
+        it { is_expected.to have_db_column :name }
+        it { is_expected.to have_db_column :email }
+        it { is_expected.to have_db_column :encrypted_password }
+    end
+
+    describe "Validations" do
+        it { is_expected.to validate_presence_of :name }
+        it { is_expected.to validate_presence_of :email }
+        it { is_expected.to validate_presence_of :encrypted_password }
```
Comment (oliverochman) [inline spec/models/user_can_create_account_spec.rb:14]: you need o test for `  validates_uniqueness_of :name` here

File: features/user_can_create_account.feature (hunk 5 above)
Comment (oliverochman) [reply features/user_can_create_account.feature]: And then remove these comments

**Review by oliverochman — CHANGES_REQUESTED** [review]

**Review by oliverochman — COMMENTED** [review]
You need to merge in development to fix the merge issues

File: features/step_definitions/user_can_create_account_assertions_steps.rb (hunk 1 above)
Comment (leiter007) [reply features/step_definitions/user_can_create_account_assertions_steps.rb]: Resolved, name changed

File: features/step_definitions/user_can_create_account_steps.rb (hunk 2 above)
Comment (leiter007) [reply features/step_definitions/user_can_create_account_steps.rb]: Resolved, name changed

File: features/step_definitions/user_can_create_account_steps.rb (hunk 3 above)
Comment (leiter007) [reply features/step_definitions/user_can_create_account_steps.rb]: Done, removed and changed after feature test changed

File: features/user_can_create_account.feature (hunk 4 above)
Comment (leiter007) [reply features/user_can_create_account.feature]: Done

File: features/user_can_create_account.feature (hunk 5 above)
Comment (leiter007) [reply features/user_can_create_account.feature]: Done as well, created for name and email uniqueness with green tests

File: spec/models/user_can_create_account_spec.rb (hunk 6 above)
Comment (leiter007) [reply spec/models/user_can_create_account_spec.rb:14]: Done as well

**Review by faraznaeem — APPROVED** [review]
The Parts of the sad path should be refactored into a background step.
But will let it slide for this time

**Pull request merged**